File: coauthor/latex/extract_figure.py
import os
import re

from ..logger import logger
from ..utils.file import readFile


def _parse_graphicspath(content: str) -> list[str]:
    r"""Parse \graphicspath commands supporting both single and multiple path formats.

    Handles both:
    \graphicspath{ {./images1/} }
    \graphicspath{ {./images1/}{./images2/} }
    """
    paths = []
    # Match both single and multiple path formats
    graphicspath_pattern = re.compile(r"\\graphicspath\s*\{((?:\s*\{[^{}]+\}\s*)+)\}")
    # Pattern to extract individual paths from nested braces
    path_pattern = re.compile(r"\{([^{}]+)\}")

    for outer_match in graphicspath_pattern.finditer(content):
        outer_content = outer_match.group(1)
        for path_match in path_pattern.finditer(outer_content):
            path = path_match.group(1).strip()
            # Ensure path has trailing slash
            if path and not path.endswith("/"):
                path += "/"
            if path:
                paths.append(path)

    return paths


def extract_figurePaths_from_latex(latexFile: str) -> list[str] | None:
    r"""Extract absolute paths of figures referenced by \includegraphics commands in LaTeX file."""
    if not latexFile or not os.path.exists(latexFile):
        return None

    content = readFile(latexFile)
    if not content:
        return None

    base_dir = os.path.dirname(latexFile)
    graphicspaths = [base_dir]  # Start with the directory of the LaTeX file

    # Parse graphicspaths
    paths = _parse_graphicspath(content)
    for path in paths:
        normalized_path = os.path.normpath(os.path.join(base_dir, path.strip("/")))
        graphicspaths.append(normalized_path)
        logger.debug(f"Added graphicspath: {normalized_path}")

    # Regular expressions to match figure inclusion commands
    figure_patterns = [r"\\includegraphics(?:\[.*?\])?\{(.*?)\}", r"\\begin\{overpic\}(?:\[.*?\])?\{(.*?)\}"]

    figurePaths = []
    for pattern in figure_patterns:
        matches = re.findall(pattern, content)
        for path in matches:
            # Try each graphics path
            for graphics_path in graphicspaths:
                fullPath_base = os.path.join(graphics_path, path)

                # Handle paths with and without extensions
                if not os.path.splitext(path)[1]:
                    for ext in [".pdf", ".png", ".jpg", ".jpeg"]:
                        fullPath = fullPath_base + ext
                        if os.path.exists(fullPath):
                            figurePaths.append(fullPath)
                            break
                else:
                    if os.path.exists(fullPath_base):
                        figurePaths.append(fullPath_base)
                        break

    if figurePaths:
        logger.info(f"Found {len(figurePaths)} figures in {latexFile}")
        return figurePaths
    return None

File: coauthor-for-vs-code/src/latex/extractFigure.ts
// Standard library imports
import * as path from 'path';

// Local imports - log
import * as logger from '../logger/logUtils';

// Local imports - utilities
import { readFile, fileExists } from '../utils/fileUtils';

const CHANNEL = 'LaTeX';
logger.initialize(CHANNEL);

/**
 * Parse graphicspath commands supporting both single and multiple path formats
 * @param content LaTeX file content
 * @returns Array of paths found in graphicspath commands
 */
function parseGraphicspath(content: string): string[] {
  const paths: string[] = [];
  // Match both single and multiple path formats
  const graphicspathPattern = /\\graphicspath\s*\{((?:\s*\{[^{}]+\}\s*)+)\}/g;
  // Pattern to extract individual paths from nested braces
  const pathPattern = /\{([^{}]+)\}/g;

  let outerMatch;
  while ((outerMatch = graphicspathPattern.exec(content)) !== null) {
    const outerContent = outerMatch[1];
    let pathMatch;
    while ((pathMatch = pathPattern.exec(outerContent)) !== null) {
      let path = pathMatch[1].trim();
      // Ensure path has trailing slash
      if (path && !path.endsWith('/')) {
        path += '/';
      }
      if (path) {
        paths.push(path);
      }
    }
  }

  return paths;
}

/**
 * Extract figure paths from a LaTeX file
 * @param latexFile Path to the LaTeX file
 * @returns Array of relative paths to figures
 */
export async function extractFigurePathsFromLatex(
  latexFile: string,
): Promise<string[]> {
  const figurePaths: string[] = [];

  try {
    const latexDir = path.dirname(latexFile);
    let graphicspaths = [latexDir]; // Start with the directory of the LaTeX file

    // Regular expressions to match figure inclusion commands
    const figurePatterns = [
      /\\includegraphics(?:\[.*?\])?\{(.+?)\}/g,
      /\\begin\{overpic\}(?:\[.*?\])?\{(.+?)\}/g,
    ];

    // Read file content
    const content = await readFile(latexFile);

    // Parse graphicspaths
    const paths = parseGraphicspath(content);
    for (const p of paths) {
      const normalizedPath = path.normalize(
        path.join(latexDir, p.replace(/^\/+|\/+$/g, '')),
      );
      graphicspaths.push(normalizedPath);
      logger.debug(CHANNEL, `Added graphicspath: ${normalizedPath}`);
    }

    logger.debug(CHANNEL, `Graphicspaths: ${graphicspaths.join(', ')}`);

    // Find all matches in the content for both patterns
    for (const pattern of figurePatterns) {
      let match;
      while ((match = pattern.exec(content)) !== null) {
        const figPath = match[1];
        for (const basePath of graphicspaths) {
          const normPath = path.normalize(path.join(basePath, figPath));
          // Try with common extensions if no extension is provided
          const extensions = figPath.includes('.')
            ? ['']
            : ['.pdf', '.png', '.jpg', '.jpeg'];

          for (const ext of extensions) {
            const pathToCheck = normPath + ext;
            const relPath = path.relative(latexDir, pathToCheck);
            if (await fileExists(relPath)) {
              figurePaths.push(relPath);
              logger.debug(CHANNEL, `Found figure: ${relPath}`);
              break;
            }
          }
        }
      }
    }

    logger.debug(CHANNEL, `Found figures: ${figurePaths.join(', ')}`);
    return figurePaths;
  } catch (err) {
    logger.error(
      CHANNEL,
      `Error extracting figure paths: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}
